fix: (storagetransfer) transfer path can be changed to empty string

gurusai-voleti opened the pull request:
<!--
Complete the self-review checklist to help speed up the review process: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

If your PR is still work in progress, please create it in draft mode.

Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to.
For example: Fixes https://github.com/hashicorp/terraform-provider-google/issues/ISSUE_ID
-->

**Release Note Template for Downstream PRs (will be copied)**
Fixes: https://github.com/hashicorp/terraform-provider-google/issues/13673
See [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) for guidance.

```release-note:bug
storagetransfer: fix to allow empty string for path
```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-14300-old..auto-pr-14300) ( 3 files changed, 206 insertions(+), 7 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-14300-old..auto-pr-14300) ( 3 files changed, 206 insertions(+), 7 deletions(-))

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccStorageTransferJob_transferPathError` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14300/artifacts/c2811db4-b557-4e29-a237-c931d761e519/recording/TestAccStorageTransferJob_transferPathError.log)]

🟢 No issues found for passed tests after REPLAYING rerun.

---






🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14300/artifacts/c2811db4-b557-4e29-a237-c931d761e519/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-14300/artifacts/c2811db4-b557-4e29-a237-c931d761e519/recording) for each test

ScottSuarez commented:
This appears to be a breaking change, see https://github.com/GoogleCloudPlatform/magic-modules/pull/14300#issuecomment-2984189491

gurusai-voleti commented:
> Hi @gurusai-voleti, what is the solution that you are proposing here?

firstly the field is missing validation, added validation to the path field as gcs object path cannot be started with "/" (reference here https://cloud.google.com/storage-transfer/docs/reference/rest/v1/GcsData) and field is an user provided and cannot be marked as computed so changed path to optional field only

as the customer reported issue they used "/" first and later realized to change to "" empty string
which they cannot as in update d.HasChanges will not detect empty string as a change and an empty string cannot be considered for computed field as per terraform sdk (error occurred with computed set to true **_A computed value with the empty string as the new value and a non-empty old value was found. Interpreting the empty string as "unset" to align with legacy behavior._**) so tf plan/apply will show no changes required, so removed computed for the field and as it is user provided no need of computed as well

as d.HasChanges and computed cannot consider empty string as a change or input, added DiffSuppressFunc if the old and new is same it will not show any difference in tf plan, if its different it will show difference in tf plan and path can be updated to "" empty string

gurusai-voleti commented:
> This appears to be a breaking change, see [#14300 (comment)](https://github.com/GoogleCloudPlatform/magic-modules/pull/14300#issuecomment-2984189491)

yes it's a breaking change, so created an PR on top 7.0 feature branch, Anything else needs to be taken care?

kautikdk commented:
If the field is currently Computed, do we know why it was marked as such? i.e Was there any change from the API side which mandated this field to get server side value as well if not provided by user?

kautikdk commented on mmv1/third_party/terraform/website/docs/guides/version_7_upgrade.html.markdown:
```diff
@@ -102,8 +102,8 @@ Description of the change and how users should adjust their configuration (if ne
+## Resource: `google_storage_transfer_job`
 
-### Resource-level change example header
+### `path` does not allow strings starting with /
```
We should add full field name here including parent block names otherwise it will mislead users. Only specifying `path` will be interpreted as top level field within the `google_storage_transfer_job` resource.

Reference: https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/website/docs/guides/version_6_upgrade.html.markdown#advanced_datapath_observability_configrelay_mode-is-now-removed

gurusai-voleti replied:
updated description

gurusai-voleti commented:
> If the field is currently Computed, do we know why it was marked as such? i.e Was there any change from the API side which mandated this field to get server side value as well if not provided by user?

even if its computed or not, as its optional field  if API response has value, it can be set otherwise not

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/97c509734ccfc6fbf806242f4dd87425c516b398..03fb9c0e0c1a9b7643ffca8cea028592d4e427f5) ( 3 files changed, 123 insertions(+), 7 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/e5c71adae593adecc01ed3c126ab206977ab12ba..a9fa8363f39cee49a2bf9db98261bb107c0e19c2) ( 3 files changed, 123 insertions(+), 7 deletions(-))

## Breaking Change(s) Detected

The following breaking change(s) were detected within your pull request.
- Field `replication_spec.gcs_data_sink.path` transitioned from optional+computed to optional `google_storage_transfer_job` - [reference](https://googlecloudplatform.github.io/magic-modules/breaking-changes/breaking-changes#field-oc-to-c)
- Field `replication_spec.gcs_data_source.path` transitioned from optional+computed to optional `google_storage_transfer_job` - [reference](https://googlecloudplatform.github.io/magic-modules/breaking-changes/breaking-changes#field-oc-to-c)
- Field `transfer_spec.gcs_data_sink.path` transitioned from optional+computed to optional `google_storage_transfer_job` - [reference](https://googlecloudplatform.github.io/magic-modules/breaking-changes/breaking-changes#field-oc-to-c)
- Field `transfer_spec.gcs_data_source.path` transitioned from optional+computed to optional `google_storage_transfer_job` - [reference](https://googlecloudplatform.github.io/magic-modules/breaking-changes/breaking-changes#field-oc-to-c)

If you believe this detection to be incorrect please raise the concern with your reviewer.
If you intend to make this change you will need to wait for a [major release](https://www.terraform.io/plugin/sdkv2/best-practices/versioning#example-major-number-increments) window.
An `override-breaking-change` label can be added to allow merging.





## Errors

`google-beta` provider:
- Failed to detect subcategory in the frontmatter in file website/docs/guides/version_7_upgrade.html.markdown.

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccStorageTransferJob_transferUpdateToEmptyString` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14300/artifacts/29506b91-be64-46e1-85ac-9bb3f5a6fea7/recording/TestAccStorageTransferJob_transferUpdateToEmptyString.log)]

🟢 No issues found for passed tests after REPLAYING rerun.

---






🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14300/artifacts/29506b91-be64-46e1-85ac-9bb3f5a6fea7/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-14300/artifacts/29506b91-be64-46e1-85ac-9bb3f5a6fea7/recording) for each test

ScottSuarez commented:
> > If the field is currently Computed, do we know why it was marked as such? i.e Was there any change from the API side which mandated this field to get server side value as well if not provided by user?
> 
> even if its computed or not, as its optional field if API response has value, it can be set otherwise not

Do we have understanding of when/in what scenario server sets a value for this field independent of user input? While yes we can switch to optional here, I am looking to see how disruptive this might be for the user.

If the user needs to deploy then define this field post deployment because the server returned some value this is not an optimal experience. In such a case we should likely explicitly set. the value to empty string if valid to avoid two step apply.

ScottSuarez commented on mmv1/third_party/terraform/services/storagetransfer/resource_storage_transfer_job.go.tmpl:
```diff
@@ -695,14 +696,32 @@ func gcsDataSchema() *schema.Resource {
 				Type:        schema.TypeString,
 				Description: `Google Cloud Storage path in bucket to transfer`,
+				ValidateFunc: validatePath,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
```
May still be relevant if API alters user input

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/97c509734ccfc6fbf806242f4dd87425c516b398..91b49dc399067639c057acaaa2ecd3932a9a9d8e) ( 3 files changed, 205 insertions(+), 7 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/e5c71adae593adecc01ed3c126ab206977ab12ba..c9e7deb26ca629cc5267349d2670dc9b2606b016) ( 3 files changed, 205 insertions(+), 7 deletions(-))

## Breaking Change(s) Detected

The following breaking change(s) were detected within your pull request.
- Field `replication_spec.gcs_data_sink.path` transitioned from optional+computed to optional `google_storage_transfer_job` - [reference](https://googlecloudplatform.github.io/magic-modules/breaking-changes/breaking-changes#field-oc-to-c)
- Field `replication_spec.gcs_data_source.path` transitioned from optional+computed to optional `google_storage_transfer_job` - [reference](https://googlecloudplatform.github.io/magic-modules/breaking-changes/breaking-changes#field-oc-to-c)
- Field `transfer_spec.gcs_data_sink.path` transitioned from optional+computed to optional `google_storage_transfer_job` - [reference](https://googlecloudplatform.github.io/magic-modules/breaking-changes/breaking-changes#field-oc-to-c)
- Field `transfer_spec.gcs_data_source.path` transitioned from optional+computed to optional `google_storage_transfer_job` - [reference](https://googlecloudplatform.github.io/magic-modules/breaking-changes/breaking-changes#field-oc-to-c)

If you believe this detection to be incorrect please raise the concern with your reviewer.
If you intend to make this change you will need to wait for a [major release](https://www.terraform.io/plugin/sdkv2/best-practices/versioning#example-major-number-increments) window.
An `override-breaking-change` label can be added to allow merging.





## Errors

`google-beta` provider:
- Failed to detect subcategory in the frontmatter in file website/docs/guides/version_7_upgrade.html.markdown.

modular-magician commented:
#### Tests analytics
Total tests: 16
Passed tests: 16
Skipped tests: 0
Affected tests: 0

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>storagetransfer</li>

</ul>

</blockquote>
</details>
🟢 All tests passed!

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-14300/artifacts/48f964bb-4c7e-424b-9c70-0500d85b77f6/build-log/replaying_test.log)

gurusai-voleti commented on mmv1/third_party/terraform/services/storagetransfer/resource_storage_transfer_job_test.go:
```diff
@@ -2399,3 +2422,159 @@ resource "google_storage_transfer_job" "transfer_job" {
+  `, project, dataSourceBucketName, project, dataSinkBucketName, project, testTransferJobName, transferJobDescription, project)
+}
+
+func testAccStorageTransferJob_transferJobGcsPath(project string, dataSourceBucketName string, dataSinkBucketName string, transferJobDescription string, testTransferJobName string, gcsPath string) string {
```
updated test case

gurusai-voleti commented:
> why it was marke



> > > If the field is currently Computed, do we know why it was marked as such? i.e Was there any change from the API side which mandated this field to get server side value as well if not provided by user?
> > 
> > 
> > even if its computed or not, as its optional field if API response has value, it can be set otherwise not
> 
> Do we have understanding of when/in what scenario server sets a value for this field independent of user input? While yes we can switch to optional here, I am looking to see how disruptive this might be for the user.
> 
> If the user needs to deploy then define this field post deployment because the server returned some value this is not an optimal experience. In such a case we should likely explicitly set the value to empty string if valid to avoid two step apply. As long as such a change isn't very divergent from normal user flow.

if the user doesn't provide any value, API doesn't set any value, it just reflects what user provided
